bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -2255,42 +2263,46 @@ private static boolean hasFillerValue(BType type, List<BType> unanalyzedTypes) {
                 BTupleType tupleType = (BTupleType) type;
                 return tupleType.getTupleTypes().stream().allMatch(TypeChecker::hasFillerValue);
             case TypeTags.UNION_TAG:
-                return checkFillerValue((BUnionType) type);
+                return checkFillerValue((BUnionType) type, unanalyzedTypes);
             default:
                 return false;
         }
     }
 
-    private static boolean checkFillerValue(BUnionType type) {
-        // NIL is a member.
-        if (type.isNullable()) {
-            return true;
-        }
-        // All members are of same type.
-        Iterator<BType> iterator = type.getMemberTypes().iterator();
-        BType firstMember;
-        for (firstMember = iterator.next(); iterator.hasNext(); ) {
-            if (!isSameType(firstMember, iterator.next())) {
-                return false;
-            }
-        }
-        // Control reaching this point means there is only one type in the union.
-        return BTypes.isValueType(firstMember) && hasFillerValue(firstMember);
-    }
-
-    private static boolean checkFillerValue(BRecordType type, List<BType> unAnalyzedTypes) {
+    private static boolean checkFillerValue(BType type, List<BType> unAnalyzedTypes) {
         if (unAnalyzedTypes.contains(type)) {
             return true;
         }
         unAnalyzedTypes.add(type);
-        for (BField field : type.getFields().values()) {
-            if (Flags.isFlagOn(field.flags, Flags.OPTIONAL)) {
-                continue;
-            }
-            if (!Flags.isFlagOn(field.flags, Flags.REQUIRED)) {
-                continue;
-            }
-            return false;
+        switch (type.getTag()) {
+            case TypeTags.RECORD_TYPE_TAG:
+                BRecordType recordType = (BRecordType) type;
+                for (BField field : recordType.getFields().values()) {
+                    if (Flags.isFlagOn(field.flags, Flags.OPTIONAL)) {
+                        continue;
+                    }
+                    if (!Flags.isFlagOn(field.flags, Flags.REQUIRED)) {
+                        continue;
+                    }
+                    return false;
+                }
+                break;
+            case TypeTags.UNION_TAG:
+                BUnionType unionType = (BUnionType) type;
+                // NIL is a member.
+                if (unionType.isNullable()) {
+                    return true;
+                }
+                // All members are of same type.
+                Iterator<BType> iterator = unionType.getMemberTypes().iterator();
+                BType firstMember;
+                for (firstMember = iterator.next(); iterator.hasNext(); ) {
+                    if (!isSameType(firstMember, iterator.next())) {
+                        return false;
+                    }
+                }
```

Comment:
Will this happen at runtime though? IINM, at compile time we get rid of duplicate types. e.g.,
```ballerina
int|int x = 10;
```
In the above union, we'll only have `int` once.

Reply:
I guess it could happen with unions of finite types with value spaces of the same basic type.

e.g.,
```ballerina
type Foo 1|2;
type Bar 0|4;

public function main() {
    (Foo|Bar)[] x = [];
    x[1] = 1;
}
```

The filler value check for `Foo|Bar` will reach here.

Reply:
Came across a separate issue when trying this out btw https://github.com/ballerina-platform/ballerina-lang/issues/24211